opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java

```diff
@@ -6,82 +6,248 @@
+      case Date: exprCoreType = ExprCoreType.TIMESTAMP;
+        break;
+      case Object: exprCoreType = ExprCoreType.STRUCT;
+        break;
+      case Nested: exprCoreType = ExprCoreType.ARRAY;
+        break;
+      case GeoPoint: return new OpenSearchGeoPointType();
+      case Binary: return new OpenSearchBinaryType();
+      case Ip: return new OpenSearchIpType();
+      default:
+        throw new IllegalArgumentException(mappingType.toString());
+    }
+    var res = new OpenSearchDataType(mappingType);
+    res.exprCoreType = exprCoreType;
+    return res;
+  }
 
   /**
-   * Parent of current type.
+   * A constructor function which builds proper `OpenSearchDataType` for given mapping `Type`.
+   * Designed to be called by the mapping parser only (and tests).
+   * @param mappingType A mapping type.
+   * @param properties Properties to set.
+   * @param fields Fields to set.
+   * @return An instance or inheritor of `OpenSearchDataType`.
    */
-  private final List<ExprType> parents;
+  public static OpenSearchDataType of(MappingType mappingType,
+                                      Map<String, OpenSearchDataType> properties,
+                                      Map<String, OpenSearchDataType> fields) {
+    var res = of(mappingType);
+    res.properties = ImmutableMap.copyOf(properties);
+    res.fields = ImmutableMap.copyOf(fields);
+    return res;
+  }
+
+  protected OpenSearchDataType(MappingType mappingType) {
+    this.mappingType = mappingType;
+  }
+
   /**
-   * JDBC type name.
+   * A constructor function which builds proper `OpenSearchDataType` for given {@link ExprType}.
+   * @param type A type.
+   * @return An instance of `OpenSearchDataType`.
    */
-  private final String jdbcType;
+  public static OpenSearchDataType of(ExprType type) {
+    if (type instanceof OpenSearchDataType) {
+      return (OpenSearchDataType) type;
+    }
+    return new OpenSearchDataType((ExprCoreType) type);
+  }
 
-  @Override
-  public List<ExprType> getParent() {
-    return parents;
+  protected OpenSearchDataType(ExprCoreType type) {
+    this.exprCoreType = type;
+  }
+
+  protected OpenSearchDataType() {
   }
 
+  // object and nested types have properties - info about embedded types
+  // a read-only collection
+  @Getter
+  @EqualsAndHashCode.Exclude
+  Map<String, OpenSearchDataType> properties = ImmutableMap.of();
+
+  // text could have fields
+  // a read-only collection
+  @EqualsAndHashCode.Exclude
+  Map<String, OpenSearchDataType> fields = ImmutableMap.of();
+
   @Override
   public String typeName() {
-    return jdbcType;
+    // To avoid breaking changes return `string` for `typeName` call (PPL) and `text` for
+    // `legacyTypeName` call (SQL). See more: https://github.com/opensearch-project/sql/issues/1296
+    if (legacyTypeName().equals("text")) {
+      return "string";
+    }
+    return legacyTypeName();
   }
 
   @Override
   public String legacyTypeName() {
```

Comment:
Please add a comment for legacyTypeName() and how it differs from ppl/legacyTypeName().
It's not obvious

Reply:
Fixed in 6e96daea0.